### PR TITLE
roachtest: fix drop test and add disk space check

### PR DESCRIPTION
**Pull request by m-schneider**
Before the drop test would fail trying to delete a large number of rows
in one transaction due to #24041.

Now we split the deletes into multiple statements. We're also checking
disk space to make sure it is cleared after all of the tpcc tables are
dropped.

Release note: None

### CODE REVIEW

File: pkg/cmd/roachtest/drop.go [hunk 1]
```diff
@@ -58,8 +62,38 @@ func init() {
...
+				t.Fatalf("failed to get range count: %v", err)
+			}
+
+			var maxWarehouse int
+			if err := db.QueryRow("select max(s_w_id) from tpcc.stock").Scan(&maxWarehouse); err != nil {
+				t.Fatalf("failed to get range count: %v", err)
+			}
+
+			var rows int
+			if err := db.QueryRow("select count(*) from tpcc.stock").Scan(&rows); err != nil {
```
Comment (tschottdorf) [inline pkg/cmd/roachtest/drop.go]: Ditto

File: pkg/cmd/roachtest/drop.go [hunk 2]
```diff
@@ -74,7 +108,43 @@ gc:
...
+					}
+
+					str := string(out)
+
+					// We need this check because sometimes the first line of the roachprod output is a warning
+					// about adding an ip to a list of known hosts.
+					if strings.Contains(str, "Warning") {
+						str = strings.Split(str, "\n")[1]
+					}
+					result := strings.Split(str, "/")
```
Comment (tschottdorf) [inline pkg/cmd/roachtest/drop.go]: I thought you would just split by whitespace. The output is something like the below, right? You just get lucky that your store-dir is absolute.

```
7070616	some/path/cockroach
```

File: pkg/cmd/roachtest/drop.go [hunk 3]
```diff
@@ -74,7 +108,43 @@ gc:
 '`
 			run(stmtZone)
 
-			// TODO(tschottdorf): assert that the disk usage drops to "near nothing".
+			// We're waiting a maximum of 5 minutes to makes sure that the drop operations clear the disk.
+			for i := 0; i < 5; i++ {
+				allNodesSpaceCleared := true
+				for j := 1; j <= nodes; j++ {
+					out, err := c.RunWithBuffer(ctx, c.l, j, fmt.Sprintf("du -sc {store-dir:%d}", j))
```
Comment (tschottdorf) [inline pkg/cmd/roachtest/drop.go]: Why `-c`? I thought `du -s` was the right one.

And I think you could cheat a little and just run

`du -s {store-dir:%d} | grep -oE '^[0-9]+`

and not bother with all the byte extraction below.

File: pkg/cmd/roachtest/drop.go [hunk 4]
```diff
@@ -74,7 +108,43 @@ gc:
...
+					if err != nil {
+						return err
+					}
+
+					str := string(out)
+
+					// We need this check because sometimes the first line of the roachprod output is a warning
+					// about adding an ip to a list of known hosts.
+					if strings.Contains(str, "Warning") {
```
Comment (tschottdorf) [inline pkg/cmd/roachtest/drop.go]: This might also just go away if you use the `grep` 😄

File: pkg/cmd/roachtest/drop.go [hunk 5]
```diff
@@ -74,7 +108,43 @@ gc:
...
+
+					size, err := strconv.Atoi(strings.TrimSpace(result[0]))
+					if err != nil {
+						return err
+					}
+
+					c.l.printf("Node %d space used: %s\n", j, humanizeutil.IBytes(int64(size)))
+
+					// Return if the size of the directory is less than 100mb
+					if size > 10000000 {
```
Comment (tschottdorf) [inline pkg/cmd/roachtest/drop.go]: could you make that `100 * 1E20`? That way we don't have to count the zeroes.

**Review by tbg — COMMENTED** [review]
Looks good! Just a few comments left.

**Comment by m-schneider** [comment]
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/cluster.go, line 654 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tl-r1-654:-L88vxrfQwQ3QoSNmiZQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/cluster.go#L654)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/drop.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-71:-L8DGa43fQQo6wocbK36:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L71)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can't you combine these two queries into one. `SELECT min(s_w_id), max(s_w_id) from tpcc.stock` and `Scan(&minWarehouse, &maxWarehouse)`.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/drop.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-76:-L88wNUGHnUKLjMcKdSY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L76)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/drop.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-115:-L88wyEbPYutIMXQMild:bdcv6kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L115)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Why `-c`? I thought `du -s` was the right one.

And I think you could cheat a little and just run

`du -s {store-dir:%d} | grep -oE '^[0-9]+`

and not bother with all the byte extraction below.
</blockquote></details>

Good point!

---

*[pkg/cmd/roachtest/drop.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-124:-L88x-6CSTFKbwW90WMe:bt902vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L124)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This might also just go away if you use the `grep` 😄 
</blockquote></details>

The warning message has an ip address, so I think I have to keep this.

---

*[pkg/cmd/roachtest/drop.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-127:-L88vShrF5wdEIPE7H-C:b-n94sih) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L127)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I thought you would just split by whitespace. The output is something like the below, right? You just get lucky that your store-dir is absolute.

```
7070616	some/path/cockroach
```
</blockquote></details>

Don't need that anymore with the grep command.

---

*[pkg/cmd/roachtest/drop.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-137:-L88xKQnh78ooj945sV1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L137)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

could you make that `100 * 1E20`? That way we don't have to count the zeroes.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081)*

**Comment by tschottdorf** [comment]
:lgtm: with remaining comments!

---

Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/cluster.go, line 654 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tl-r1-654:-L8DLhrbgnjCi3mGC2p2:b38v8th) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/cluster.go#L654)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not seeing it, did you push?

---

*[pkg/cmd/roachtest/drop.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-124:-L8DLB2EfME6RyMnwCo9:bc9eqw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L124)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

The warning message has an ip address, so I think I have to keep this.
</blockquote></details>

No, the grep is anchored at the beginning of the line. Give it a try!

---

*[pkg/cmd/roachtest/drop.go, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L8DLUdxzuAaKewxiMLb:-L8DLUdzvL8ZLjYPDljx:b1qqyro) ([raw file](https://github.com/cockroachdb/cockroach/blob/097af46dd4cc486d6812a5a33502412d29fb99f0/pkg/cmd/roachtest/drop.go#L106)):*
> ```Go
> 				allNodesSpaceCleared := true
> 				for j := 1; j <= nodes; j++ {
> 					out, err := c.RunWithBuffer(ctx, c.l, j, fmt.Sprintf("du -s {store-dir:%d} | grep -oE '^[0-9]+'", j))
> ```

Just as a sanity check, could you also run this before starting to delete stuff and verify that it's at least <X> (not sure how big X is but just look at the initial data set and take half of that)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-:-L8DLf1rsmTrpg_R6myt:bto2xp4)*

**Comment by m-schneider** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/cmd/roachtest/cluster.go, line 654 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tl-r1-654:-L8DeKL-XAix-572PQ0w:b-q32h51) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/cluster.go#L654)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not seeing it, did you push?
</blockquote></details>

Ah yes.

---

*[pkg/cmd/roachtest/drop.go, line 124 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-124:-L8DeR43WTGtFojbvfPX:b-4x5hq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L124)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No, the grep is anchored at the beginning of the line. Give it a try!
</blockquote></details>

I did and it broke.

---

*[pkg/cmd/roachtest/drop.go, line 106 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24081#-L8DLUdxzuAaKewxiMLb:-L8EJwX7QU5E_KmqGicf:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/097af46dd4cc486d6812a5a33502412d29fb99f0/pkg/cmd/roachtest/drop.go#L106)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Just as a sanity check, could you also run this before starting to delete stuff and verify that it's at least <X> (not sure how big X is but just look at the initial data set and take half of that)?
</blockquote></details>

Added

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24081)*

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/cmd/roachtest/drop.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081#-L842PpDmb-uEwasD8Tm-r1-124:-L8F11JGB0p-JDIo3Jwq:b-4bjoi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf31e443fa9e9a508868b5499a8c0677c9b7c136/pkg/cmd/roachtest/drop.go#L124)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I did and it broke.
</blockquote></details>

Oh right, the line is output before running the command :see_no_evil: We should probably just suppress that warning in general, but not in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24081)*

**Pull request merged**